### PR TITLE
Improve jdk download tests

### LINKED ISSUE #41998: JdkDownloadPluginIT testDarwinExtraction/testWindowsExtraction failures
```
* What went wrong:
Could not determine the dependencies of task ':extractDarwinJdk1.0.2+99'.
> Could not resolve all files for configuration ':openjdk_1.0.2+99_darwin'.
   > Could not resolve jdk:osx:1.0.2.
     Required by:
         project :
      > Could not resolve jdk:osx:1.0.2.
         > Could not get resource 'https://download.oracle.com/java/GA/jdk1.0.2/null/99/GPL/openjdk-1.0.2_osx-x64_bin.tar.gz'.
            > Could not HEAD 'https://download.oracle.com/java/GA/jdk1.0.2/null/99/GPL/openjdk-1.0.2_osx-x64_bin.tar.gz'.
               > Read timed out
```

Note the suspicious **null** in the URL and the odd JDK version (1.0.2?). 

https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+7.x+multijob-unix-compatibility/os=sles-12/157/console

### PR DESCRIPTION
**Pull request by rjernst**
This commit reworks the tests for jdk download to test the old and new
url pattern from oracle. Additionally it limits to one repository
created per version, based on the old or new pattern, and restricts
other repositories from trying to resolve jdks.

closes #41998

### CODE REVIEW

File: buildSrc/src/main/java/org/elasticsearch/gradle/JdkDownloadPlugin.java [hunk 1]
```diff
@@ -95,24 +96,34 @@ private static void setupRootJdkDownload(Project rootProject, String platform, S
...
-            });
-            // current pattern since 12.0.1
-            rootProject.getRepositories().ivy(ivyRepo -> {
-                ivyRepo.setName(repoName + "_with_hash");
-                ivyRepo.setUrl("https://download.oracle.com");
-                ivyRepo.metadataSources(IvyArtifactRepository.MetadataSources::artifact);
-                ivyRepo.patternLayout(layout -> layout.artifact(
-                    "java/GA/jdk" + jdkVersion + "/" + hash + "/" + jdkBuild + "/GPL/openjdk-[revision]_[module]-x64_bin.[ext]"));
-                ivyRepo.content(content -> content.includeGroup("jdk"));
+            if (Integer.parseInt(jdkMajor) >= 12 && jdkVersion.equals("12") == false) {
```
Comment (mark-vieira) [inline buildSrc/src/main/java/org/elasticsearch/gradle/JdkDownloadPlugin.java]: Just so I'm following, a `jdkVersion` here would be "12" in the case of 12.0.0 and I guess "12.0.1" for the new pattern?

Comment (mark-vieira) [reply buildSrc/src/main/java/org/elasticsearch/gradle/JdkDownloadPlugin.java]: Alternatively could we not just do `if (hash != null)`?

File: buildSrc/src/test/java/org/elasticsearch/gradle/JdkDownloadPluginIT.java [hunk 2]
```diff
@@ -41,60 +41,75 @@
...
     }
 
     public void testDarwinExtraction() throws IOException {
-        assertExtraction("getDarwinJdk", "osx", "Contents/Home/bin/java");
+        assertExtraction("getDarwinJdk", "osx", "Contents/Home/bin/java", JDK_VERSION);
     }
 
     public void testWindowsExtraction() throws IOException {
-        assertExtraction("getWindowsJdk", "windows", "bin/java");
+        assertExtraction("getWindowsJdk", "windows", "bin/java", JDK_VERSION);
```
Comment (mark-vieira) [inline buildSrc/src/test/java/org/elasticsearch/gradle/JdkDownloadPluginIT.java]: FWIW, this kind of stuff if where Spock data tables are very convenient 😉

File: buildSrc/src/main/java/org/elasticsearch/gradle/JdkDownloadPlugin.java [hunk 3]
```diff
@@ -95,24 +96,34 @@ private static void setupRootJdkDownload(Project rootProject, String platform, S
...
+                    ivyRepo.setName(repoName);
+                    ivyRepo.setUrl("https://download.oracle.com");
+                    ivyRepo.metadataSources(IvyArtifactRepository.MetadataSources::artifact);
+                    ivyRepo.patternLayout(layout ->
+                        layout.artifact("java/GA/jdk" + jdkMajor + "/" + jdkBuild + "/GPL/openjdk-[revision]_[module]-x64_bin.[ext]"));
+                    ivyRepo.content(content -> content.includeGroup("jdk"));
+                });
+            }
+            // all other repos should ignore the special jdk artifacts
+            repositories.all(repo -> {
```
Comment (mark-vieira) [inline buildSrc/src/main/java/org/elasticsearch/gradle/JdkDownloadPlugin.java]: By having this inside the jdk container loop we end up iterating over the repo container and adding this exclusion multiple times. We should do this only once, outside the loop and simply check that the repo name doesn't begin with `jdk_repo_`.

**Review by mark-vieira — APPROVED** [review]
Couple questions and comments but otherwise LGTM 👍

File: buildSrc/src/main/java/org/elasticsearch/gradle/JdkDownloadPlugin.java (hunk 1 above)
Comment (rjernst) [reply buildSrc/src/main/java/org/elasticsearch/gradle/JdkDownloadPlugin.java]: Yes that is correct. You are right we could just check for the hash. I will change to that.

**Comment by rjernst** [comment]
@elasticmachine run elasticsearch-ci/1

**Pull request merged**